pkg/procs/clone.go

```diff
@@ -0,0 +1,120 @@
+/*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+* Copyright 2021 Red Hat, Inc.
+ */
+
+package procs
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strconv"
+)
+
+// the proc layout we care about is more volatile than the ghw/snapshot package expects to deal with. So we need extra care.
+func CloneInto(logger *log.Logger, procfsRoot, scratchDir string) error {
+	baseDir := filepath.Join(scratchDir, "proc")
+
+	var err error
+	if err = os.MkdirAll(baseDir, os.ModePerm); err != nil {
+		return err
+	}
+
+	pidEntries, err := ioutil.ReadDir(procfsRoot)
+	if err != nil {
+		return err
+	}
+	for _, pidEntry := range pidEntries {
+		if !pidEntry.IsDir() {
+			continue
+		}
+		pid, err := strconv.Atoi(pidEntry.Name())
+		if err != nil {
+			// doesn't look like a pid
+			continue
+		}
+
+		err = CloneEntry(logger, pid, procfsRoot, scratchDir)
+		if err != nil {
+			logger.Printf("cannot clone data for pid %d (skipped): %v", pid, err)
+			continue
+		}
+	}
+	return nil
+}
+
+func CloneEntry(logger *log.Logger, pid int, procfsRoot, scratchDir string) error {
```

Comment:
I think the `NewSnapProcCommand` description and the real functionality is a little bit different, I excepted it will snapshot all files under the `/proc/*/`(maybe except files under the `fd`) and not only `status` and `cmdline`.

Reply:
Thing is snapshotting the whole `/proc` tree is not trivial. So we start actually collecting what we need we use. It's debatable indeed, but it's all about tradeoffs.